compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1132,10 +1132,13 @@ class Typer extends Namer
         // if expected result type is a wildcard, approximate from above.
         // this can type the greatest set of admissible closures.
         (pt1.argTypesLo.init, typeTree(interpolateWildcards(pt1.argTypesHi.last)))
-      case SAMType(sam @ MethodTpe(_, formals, restpe)) =>
+      case RefinedType(parent, nme.apply, mt @ MethodTpe(_, formals, restpe))
+      if defn.isNonRefinedFunction(parent) && formals.length == defaultArity =>
+        (formals, untpd.DependentTypeTree(syms => restpe.substParams(mt, syms.map(_.termRef))))
```

Comment:
Do we have a testcase for this change?

Reply:
In fact I could not come up with one. The only test that fails in the SAM case was scala/reflect/TypeTest.scala. Making the analogous example based on dependent functions instead of SAMs passed even without this patch. I am going to include that test. I have reverted the change, since it's not clear we need it.